ssh-agent.c

```diff
@@ -461,31 +600,47 @@ process_sign_request2(SocketEntry *e)
 		verbose_f("%s key not found", sshkey_type(key));
 		goto send;
 	}
-	/*
-	 * If session IDs were recorded for this socket, then use them to
-	 * annotate the confirmation messages with the host keys.
-	 */
-	if (e->nsession_ids > 0 &&
-	    parse_userauth_request(data, key, &user, &sid) == 0) {
-		for (i = 0; i < e->nsession_ids; i++) {
-			if ((fp = sshkey_fingerprint(e->session_ids[i].key,
-			    SSH_FP_HASH_DEFAULT, SSH_FP_DEFAULT)) == NULL)
-				fatal_f("fingerprint failed");
-			if (buf_equal(sid, e->session_ids[i].sid) == 0) {
-				debug3_f("destination %s %s (slot %zu)",
-				    sshkey_type(e->session_ids[i].key), fp, i);
-				xasprintf(&sig_dest, "public key request for "
-				    "user \"%s\" to %s %s", user,
-				    sshkey_type(e->session_ids[i].key), fp);
-				free(fp);
-				fp = NULL;
-				break;
-			}
-			debug3_f("forwarded by %s %s (slot %zu)",
-			    sshkey_type(e->session_ids[i].key), fp, i);
-			free(fp);
-			fp = NULL;
+	if ((fp = sshkey_fingerprint(key, SSH_FP_HASH_DEFAULT,
+	    SSH_FP_DEFAULT)) == NULL)
+		fatal_f("fingerprint failed");
+
+	if (id->ndest_constraints != 0) {
+		if (e->nsession_ids == 0) {
+			logit_f("refusing use of destination-constrained key "
+			    "to sign on unbound connection");
+			goto send;
+		}
+		if (parse_userauth_request(data, key, &user, &sid) != 0) {
+			logit_f("refusing use of destination-constrained key "
+			   "to sign unidentified signature");
+			goto send;
 		}
+		/* XXX logspam, leak */
+		debug_f("user=%s sid=%s", user,
+		    tohex(sshbuf_ptr(sid), sshbuf_len(sid)));
+		if (check_destination_constraints(id, e, user, &hostname) != 0)
+			goto send;
+		/*
+		 * Ensure that the session ID is the most recent one
+		 * registered on the socket - it should have been bound by
+		 * ssh immediately before userauth.
+		 */
```

Comment:
when would there be multiple entries? only for
ssh -A host1 ssh -A host2 ssh host3 ?

Reply:
also just "ssh -A host1 ssh host2", because the session ID/hostkey binding gets passed in pubkey_prepare() too